source/jsonParsing.ml

```diff
@@ -7,60 +7,280 @@
+  let string_member ?default name json = member name json |> to_string_with_default ?default
 
-let optional_member ~f name json =
-  member name json
-  |> function
-  | `Null -> None
-  | _ as element -> Some (f element)
+  let optional_member ~f name json =
+    member name json
+    |> function
+    | `Null -> None
+    | _ as element -> Some (f element)
 
 
-let optional_string_member = optional_member ~f:to_string
+  let optional_string_member = optional_member ~f:to_string
 
-let optional_int_member = optional_member ~f:to_int
+  let optional_int_member = optional_member ~f:to_int
 
-let path_member name json = member name json |> to_path
+  let path_member name json = member name json |> to_path
 
-let optional_path_member = optional_member ~f:to_path
+  let optional_path_member = optional_member ~f:to_path
 
-let list_member ?default ~f name json =
-  member name json
-  |> fun element ->
-  match element, default with
-  | `Null, Some default -> default
-  | _, _ -> convert_each f element
+  let list_member ?default ~f name json =
+    member name json
+    |> fun element ->
+    match element, default with
+    | `Null, Some default -> default
+    | _, _ -> convert_each f element
 
 
-let optional_list_member ~f name json =
-  member name json
-  |> function
-  | `Null -> None
-  | element -> Some (convert_each f element)
+  let optional_list_member ~f name json =
+    member name json
+    |> function
+    | `Null -> None
+    | element -> Some (convert_each f element)
 
 
-let string_list_member = list_member ~f:to_string
+  let string_list_member = list_member ~f:to_string
 
-let path_list_member = list_member ~f:to_path
+  let path_list_member = list_member ~f:to_path
+end
+
+module JsonAst = struct
+  open Core
+  module Result = Core.Result
+
+  module Location = struct
+    type position = {
+      line: int;
+      column: int;
+    }
+
+    and t = {
+      start: position;
+      stop: position;
+    }
+    [@@deriving show, eq]
+
+    let null_position = { line = -1; column = -1 }
+
+    let null_location = { start = null_position; stop = null_position }
+
+    let pp_start formatter { start = { line; column }; _ } =
+      Format.fprintf formatter "%d:%d" line column
+
+
+    let from_decoded_range range =
+      let (start_line, start_column), (end_line, end_column) = range in
+      {
+        start = { line = start_line; column = start_column };
+        stop = { line = end_line; column = end_column };
+      }
+  end
+
+  module Node = struct
+    type 'a t = {
+      location: Location.t;
+      value: 'a;
+    }
+    [@@deriving show, eq]
+  end
+
+  module Error = struct
+    type kind = ParseError
+
+    type t = {
+      kind: kind;
+      message: string;
+      location: Location.t;
+    }
+
+    let create ~kind ~location message =
+      match kind with
+      | ParseError -> { kind = ParseError; message; location }
+  end
```

Comment:
This is only used for parse errors, so we could make this a lot simpler:
```suggestion
  module ParseError = struct
    type t = {
      message: string;
      location: Location.t;
    }
  end
```

Reply:
Okay!